Add troubleshooting guide (#919)
Shopify/vscode-ruby-lsp: TROUBLESHOOTING.md

inject it into the NodeJS process and appropriately set the Ruby version and gem installation paths.

As an example, the activation script for `zsh` using `rbenv` as a version manager will look something like this:

```shell
# Invoke zsh using interactive mode (loads ~/.zshrc) to run a single command
# The command is `rbenv exec ruby`, which automatically sets all relevant environment variables and selects the
# specified Ruby version
# We then print the activated environment as a JSON. We read that JSON from the NodeJS process to insert the needed
# environment variables in order to run Ruby correctly
/bin/zsh -ic 'rbenv exec ruby -rjson -e "puts JSON.dump(ENV)"'
```

After activating the Ruby version, we then proceed to boot the server gem (`ruby-lsp`). To avoid having users include
the `ruby-lsp` in their `Gemfile`, we currently create a custom bundle under the `.ruby-lsp` directory inside your
project. That directory contains another `Gemfile`, that includes the `ruby-lsp` gem in addition to your project's
dependencies. This approach allows us to automatically detect which formatter your project uses and which gems we need
to index for features such as go to definition.

> **Note**: we are working with the rubygems/bundler team to have this type of mechanism properly supported from within
> Bundler itself, which is currently being experimented with in a plugin called `bundler-compose`. Once `bundler-compose`
> is production ready, the entire custom bundle created under the `.ruby-lsp` directory will go away and we'll rely on
> Bundler to compose the LOAD_PATH including the `ruby-lsp` gem.

Reviewer: If you want to use GH's special syntax for note, you need to put this section under quote too:
```suggestion
> **Note**: we are working with the rubygems/bundler team to have this type of mechanism properly supported from within
> Bundler itself, which is currently being experimented with in a plugin called `bundler-compose`. Once `bundler-compose`
> is production ready, the entire custom bundle created under the `.ruby-lsp` directory will go away and we'll rely on
> Bundler to compose the LOAD_PATH including the `ruby-lsp` gem.
```

If that's not the case, just ignore this comment 😄

environment variable. This usually ends up in having `/bin/sh` selected instead of your actual shell. If you are facing
this problem, please try to

- Update VS Code to the latest version
- Completely close VS Code and launch it from the terminal with `code .` (instead of opening VS Code from the launch
  icon)

More context about this issue on https://github.com/Shopify/vscode-ruby-lsp/issues/901.

### Bundler issues

If the extension successfully activated the Ruby environment, it may still fail when trying to compose the custom bundle
to run the server gem. This could be a regular bundler issue, like not being able to satisfy dependencies due to a
conflicting version requirement, or it could be a configuration issue.

For example, if the project has its linter/formatter put in an optional `Gemfile` group and that group is excluded in
the Bundler configuration, the Ruby LSP will not be able to see those gems.

```ruby
# Gemfile

# ...

# If Bundler is configured to exclude this group, the Ruby LSP will not be able to find `rubocop`
group :optional_group do
  gem "rubocop"
end
```

> If you experience Bundler related issues, double-check both your global and project-specific configuration to check if

Reviewer: Could we list a few file locations/commands that may help users on this? I think `bundle config --help` should be a very good first step for them to make the first step.

there's anything that could be preventing the server from booting.

### Format on save dialogue won't disappear

When VS Code requests formatting for a document, it opens a dialogue showing progress a couple of seconds after sending
the request, closing it once the server has responded with the formatting result.

If you are seeing that the dialogue is not going away, this likely doesn't mean that formatting is taking very long or
hanging. It likely means that the server crashed or got into a corrupt state and is simply not responding to any
requests, which means the dialogue will never go away.

This is always the result of a bug in the server. It should always fail gracefully without getting into a corrupt state

> that will prevent it from responding to new requests coming from the editor. If you encounter this, please submit a bug

Reviewer: I think we can shorten this sentence a bit:

```suggestion
that prevents its from responding to new requests. If you encounter this, please submit a bug
```
